pkg/flow/logging/logger.go

```diff
@@ -85,11 +92,25 @@ func (l *Logger) Update(o Options) error {
 	}
 	l.writer.Set(newWriter)
 
+	// Print out the buffered logs since we determined the log format already
+	for _, bufferedLogChunk := range l.buffer {
+		if err := slogadapter.GoKit(l.handler).Log(bufferedLogChunk...); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
 // Log implements log.Logger.
 func (l *Logger) Log(kvps ...interface{}) error {
+	// Buffer logs before confirming log format is configured in `logging` block
+	l.bufferMut.Lock()
+	defer l.bufferMut.Unlock()
```

Comment:
This adds a bottleneck to logging where only one goroutine can log at a time; I'd recommend changing `l.bufferMut` to a RWMutex and grab a read lock here instead.

Reply:
I've addressed. PTAL

Reply:
Sadly there's a [data race](https://github.com/grafana/agent/actions/runs/7786596192/job/21231781964) when I switched to read lock usage since we're growing the buffer below as well. I've reverted back to vanilla mutex

Reply:
Yep, you're right about the datarace. I still think this is a performance bottleneck though, so we need to be a little more careful about how we write this code. 

For example, something like this could work, but other solutions might be appropriate too:

```go
l.bufferMut.RLock() 
if !l.hasLogFormat {
  l.bufferMut.RUnlock() 
  l.bufferMut.Lock() 
  // Check hasLogFormat again; could have changed since the unlock. 
  if !l.hasLogFormat {
    l.buffer = append(l.buffer, kvps) 
    return nil 
  }
  l.bufferMut.Unlock() 
} else {
  l.bufferMut.RUnlock()
}
```

It's quite ugly, but it would work. If you can find a cleaner way of implementing this it would be appreciated 🙂

Reply:
there is a little bug here: there should be an `l.bufferMut.Unlock()` right before the `return nil`, else the mutex stays locked forever

Reply:
Great catch! Amended :D